pinot-spi/src/main/java/org/apache/pinot/spi/utils/retry/BaseRetryPolicy.java

```diff
@@ -42,26 +42,26 @@ protected BaseRetryPolicy(int maxNumAttempts) {
   protected abstract long getDelayMs(int currentAttempt);
 
   @Override
-  public void attempt(Callable<Boolean> operation)
+  public int attempt(Callable<Boolean> operation)
       throws AttemptsExceededException, RetriableOperationException {
     int attempt = 0;
     try {
       while (attempt < _maxNumAttempts - 1) {
         if (Boolean.TRUE.equals(operation.call())) {
           // Succeeded
-          return;
+          return attempt;
         } else {
           // Failed
           Thread.sleep(getDelayMs(attempt++));
         }
       }
       if (_maxNumAttempts > 0 && Boolean.TRUE.equals(operation.call())) {
         // Succeeded
-        return;
+        return attempt;
       }
     } catch (Exception e) {
-      throw new RetriableOperationException(e);
+      throw new RetriableOperationException(e, attempt + 1);
     }
-    throw new AttemptsExceededException("Operation failed after " + _maxNumAttempts + " attempts");
+    throw new AttemptsExceededException("Operation failed after " + _maxNumAttempts + " attempts", attempt);
```

Comment:
```suggestion
    throw new AttemptsExceededException("Operation failed after " + _maxNumAttempts + " attempts", _maxNumAttempts);
```
Should we return _maxNumAttempts here?

Reply:
done